File: 0518 - Coin Change 2/solution.py
class Solution:
    def change(self, amount: int, A: List[int]) -> int:
        @lru_cache(None)
        def dp(i, amt):
            if amt <= 0 or i == len(A):
                return int(amt == 0)
            return dp(i, amt-A[i])+dp(i+1, amt)
        A.sort(reverse=True)
        return dp(0, amount)

File: 0518 - Coin Change 2/solution.c
int change(int amount, int* A, int ASz){
    int dp[5001] = {1};
    for(int i = 0; i < ASz; i++)
        for(int j = A[i]; j <= amount; j++)
            dp[j] += dp[j-A[i]];
    return dp[amount];
}